sycl/test-e2e/InvokeSimd/Regression/nbarrier_loop.cpp

```diff
@@ -0,0 +1,212 @@
+// REQUIRES: gpu-intel-pvc && linux
+// UNSUPPORTED: cuda || hip
+//
+// TODO: enable when Jira issue resolved, currently fail with VISALTO enable
+// XFAIL: gpu-intel-pvc
+//
+// RUN: %{build} -fno-sycl-device-code-split-esimd -Xclang -fsycl-allow-func-ptr -o %t.out
+// RUN: env IGC_VCSaveStackCallLinkage=1 IGC_VCDirectCallsOnly=1 %{run} %t.out
+//
+// VISALTO enable run
+// RUN: env IGC_VISALTO=63 IGC_VCSaveStackCallLinkage=1 IGC_VCDirectCallsOnly=1 %{run} %t.out
+
+/*
+ * Test checks support of named barrier in a loop in invoke_simd context.
+ * First iteration has 1 barrier and 1 producer, second - 2 barriers and 2
+ * producers. Producer stores data to SLM, then all threads read SLM and store
+ * data to surface.
+ */
+
+#include <sycl/ext/intel/esimd.hpp>
+#include <sycl/ext/oneapi/experimental/invoke_simd.hpp>
+#include <sycl/sycl.hpp>
+
+#include <sycl/ext/intel/esimd.hpp>
+#include <sycl/ext/intel/experimental/esimd/memory.hpp>
+#include <sycl/ext/oneapi/experimental/invoke_simd.hpp>
+#include <sycl/sycl.hpp>
+
+#include <functional>
+#include <iostream>
+#include <type_traits>
+
+// TODO: When gpu driver can pass/accept accessor by value,
+// the work-around undef #ifdef US_ACC_PTR should be removed.
+#define USE_ACC_PTR
+
+using namespace sycl;
+using namespace sycl::ext::oneapi::experimental;
+namespace esimd = sycl::ext::intel::esimd;
+namespace experimental_esimd = sycl::ext::intel::experimental::esimd;
+
+constexpr int VL = 16;
+constexpr unsigned Groups = 1;
+constexpr unsigned Threads = 8;
+constexpr unsigned Size = Groups * Threads * VL;
+
+class KernelID;
+
+ESIMD_INLINE void ESIMD_CALLEE_nbarrier(local_accessor<int, 1> LocalAcc,
+                                        int localID,
+                                        int *o) SYCL_ESIMD_FUNCTION {
+  // 2 named barriers, id 0 reserved for unnamed
+  constexpr unsigned bnum = 3;
+
+  experimental_esimd::named_barrier_init<bnum>();
+
+  // slm size used in kernel
+  constexpr unsigned slm_size = Size / 2;
+
+  // 2 producers on first iteration, 1 producer on second
+  unsigned int indexes[2][2] = {{1, 2}, {3, 3}}; // local ids of producers
+  unsigned int prods[2] = {2, 1};                // number of producers
+
+  unsigned int off = localID * VL;
+  // producer writes to SLM, consumer reads what producer wrote
+  unsigned int slm_base = static_cast<uint32_t>(
+      reinterpret_cast<std::uintptr_t>(LocalAcc.get_pointer()));
+
+  esimd::barrier();
+
+  for (int b = bnum - 1; b > 0; b--) {
+    int j = bnum - b - 1; // iteration index
+
+    bool is_producer = localID == indexes[j][0] || localID == indexes[j][1];
+    bool is_consumer = !is_producer;
+    // only-consumer or only-producer modes
+    unsigned int flag = is_producer ? 0x1 : 0x2;
+
+    unsigned int producers = prods[j];
+    unsigned int consumers = Threads - producers;
+
+    if (is_producer) {
+      unsigned int slm_store_off = j * sizeof(int) * slm_size / 4;
```

Comment:
can you add some explanation to the constant division/multiplication? for ex i don't understand the /4 here.  if it's for a data type, can you use sizeof(datatype)?

Reply:
I reworked this part a bit and added some comments. Hope now it is more clear what's going on in the test.